Expand is_functional_schema to work with torch._C._FunctionSchema

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123454
* #123453
* #123437
* #123129
* #123110
* #123109
* __->__ #123108

Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

zou3519 committed:
Expand is_functional_schema to work with torch._C._FunctionSchema

Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Expand is_functional_schema to work with torch._C._FunctionSchema"


Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_library/utils.py:
```diff
@@ -78,24 +78,29 @@ def is_functional_schema(schema: Any) -> bool:
-    if not schema.returns:
-        return False
-    return True
+    return is_functional(schema)
```
nit: this can be re-ordered up by doing a single assert + if + this call.

albanD commented on torch/_C/__init__.pyi.in:
```diff
@@ -854,6 +854,7 @@ class FunctionSchema:
     returns: List[Argument]
     name: str
     overload_name: str
+    is_mutable: _bool
```
Why do we have this field and not just look it up on the arguments?

albanD replied:
Isn't that what you do in your function is_functional that you added below?

zou3519 replied:
Is your suggestion to delete `is_mutable` and add a helper function `is_mutable` to do what it previously did? If so, there already is an `is_mutable` helper function in C++ (that's what this field is).

NB: `is_mutable` and `is_functional_schema` below aren't opposites :/ we have too many variants of these checks flying around and I'm trying to not break any existing code (is_mutable exists already today).

zou3519 replied:
NB: is_mutable is a @property, so we're good.

zou3519 committed:
Update on "Expand is_functional_schema to work with torch._C._FunctionSchema"


Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Expand is_functional_schema to work with torch._C._FunctionSchema"


Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Expand is_functional_schema to work with torch._C._FunctionSchema"


Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_library/utils.py:
```diff
@@ -78,24 +78,29 @@ def is_functional_schema(schema: Any) -> bool:
+            return False
+        rets = schema.returns
+        is_non_mutating_view = len(rets) > 0 and any(
+            r.alias_info is not None and not r.alias_info.is_write for r in rets
```
How could "is_write" be true here since we check for is_mutable False above?

zou3519 replied:
is_mutable only checks that the inputs are written to :/
All this code is pre-existing, I am not changing the logic. I agree that it's messy.

albanD commented on torch/_library/utils.py:
```diff
@@ -78,24 +78,29 @@ def is_functional_schema(schema: Any) -> bool:
+        if is_non_mutating_view:
+            return False
+        if not schema.returns:
+            return False
```
this is a weird schema: not inplace and no output haha

zou3519 replied:
this usually just means the user did write an inplace op but forgot to label it as inplace.

albanD reviewed (approved):
Only nits in the logic that is weird, but most likely due to the complexity of the schema so not blocking the PR

zou3519 committed:
Update on "Expand is_functional_schema to work with torch._C._FunctionSchema"


Previously it worked with torchgen.model.FunctionSchema. This PR extends
it to work with torch._C._FunctionSchema by making
torchgen.model.FunctionSchema look more like torch._C._FunctionSchema.

Test Plan:
- new tests

[ghstack-poisoned]
